pkg/nsx/services/subnetbinding/tree.go

```diff
@@ -0,0 +1,221 @@
+	childRes := model.ChildResourceReference{
+		Id:           &resID,
+		ResourceType: "ChildResourceReference",
+		TargetType:   &targetType,
+		Children:     children,
+	}
+	dataValue, errors := common.NewConverter().ConvertToVapi(childRes, model.ChildResourceReferenceBindingType())
+	if len(errors) > 0 {
+		return nil, errors[0]
+	}
+	return []*data.StructValue{dataValue.(*data.StructValue)}, nil
+}
+
+func wrapSubnetBindingMap(bindingMap *model.SubnetConnectionBindingMap) (*data.StructValue, error) {
+	bindingMap.ResourceType = &common.ResourceTypeSubnetConnectionBindingMap
+	childBindingMap := model.ChildSubnetConnectionBindingMap{
+		Id:                         bindingMap.Id,
+		MarkedForDelete:            bindingMap.MarkedForDelete,
+		ResourceType:               "ChildSubnetConnectionBindingMap",
+		SubnetConnectionBindingMap: bindingMap,
+	}
+	dataValue, errors := common.NewConverter().ConvertToVapi(childBindingMap, model.ChildSubnetConnectionBindingMapBindingType())
+	if len(errors) > 0 {
+		return nil, errors[0]
+	}
+	return dataValue.(*data.StructValue), nil
+}
+
+func (s *BindingService) hUpdateSubnetConnectionBindingMaps(subnetPath string, bindingMaps []*model.SubnetConnectionBindingMap) error {
+	vpcInfo, err := common.ParseVPCResourcePath(subnetPath)
+	if err != nil {
+		return err
+	}
+	subnetID := vpcInfo.ID
+	orgRoot, err := buildOrgRootBySubnetConnectionBindingMaps(bindingMaps, subnetPath)
+	if err != nil {
+		return err
+	}
+
+	if err = s.NSXClient.OrgRootClient.Patch(*orgRoot, &enforceRevisionCheckParam); err != nil {
+		log.Error(err, "Failed to patch SubnetConnectionBindingMaps on NSX", "orgID", vpcInfo.OrgID, "projectID", vpcInfo.ProjectID, "vpcID", vpcInfo.VPCID, "subnetID", subnetID, "subnetConnectionBindingMaps", bindingMaps)
+		err = nsxutil.TransNSXApiError(err)
+		return err
+	}
+
+	// Get SubnetConnectionBindingMaps from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	subnetBindingListResult, err := s.NSXClient.SubnetConnectionBindingMapsClient.List(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, subnetID, nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		log.Error(err, "Failed to list SubnetConnectionBindingMaps from NSX under subnet", "orgID", vpcInfo.OrgID, "projectID", vpcInfo.ProjectID, "vpcID", vpcInfo.VPCID, "subnetID", subnetID, "subnetConnectionBindingMaps", bindingMaps)
+		err = nsxutil.TransNSXApiError(err)
+		return err
+	}
+
+	nsxBindingMaps := make(map[string]model.SubnetConnectionBindingMap)
+	for _, bm := range subnetBindingListResult.Results {
+		nsxBindingMaps[*bm.Id] = bm
+	}
+
+	for i := range bindingMaps {
+		bm := bindingMaps[i]
+		if bm.MarkedForDelete != nil && *bm.MarkedForDelete {
+			s.BindingStore.Apply(bm)
+		} else {
+			nsxBindingMap := nsxBindingMaps[*bm.Id]
+			s.BindingStore.Apply(&nsxBindingMap)
+		}
+	}
+
+	return nil
+}
+
+func (s *BindingService) hDeleteSubnetConnectionBindingMap(bindingMaps []*model.SubnetConnectionBindingMap) error {
```

Comment:
When this is used in a total cleanup, is it possible for hDeleteSubnetConnectionBindingMap to exceed the maximum transaction size of a single hierarchical API call? Do we need to split the deletion into multiple calls?

Reply:
Good catch, I would add paging logic to avoid the potential issue.